### ISSUE TITLE
Can any am architecture be used for streaming?

### ISSUE THREAD

**Issue by mjhanphd** [facebookresearch/wav2letter#617]
First of all please forgive me for asking this fundamental question before digging into the code myself.
I looked through your streaming convnet paper and codes, and I was wondering if I could simply replace the AM architecture by transformer model, e.g. just the model in your sota model (am_transformer_ctc.arch, I found that only ctc criterion allows for streaming).
Since I know that the transformer taking future contexts in its self attention layers wouldn't be feasible when it comes to 'streaming', I simply gave 'useMask = true' to eliminate any disagreement between training and decoding phases.
So my question goes here : Is this enough to make transformer AM model work in streaming fashion?

Since I'll need a little bit more of time to figure this out, I'll be so happy if anyone give me a great tip.
Thank you :)

**Comment by mjhanphd** [comment]
I've continuously been thinking about this problem during the weekend.
It seems like transformer and tdss convnet aren't very different from each other when they are seen from outside: both just deal with 'a sequence of feature vectors' (or simply, a matrix) to produce a (predicted) sequence of tokens.
One thing I really am confused and hard to find out from your code follows:
For a model/decoder to be genuinely online, the AM architecture should also work in online fashion.
For instance if A+B (A first, and B next in a certain period of time) comes in to the AM model, then the model should compute A first and then (only) B the next time, instead of A first and then A+B next.
Yet the model should leverage the information that it obtained from A to better encode B.
-> Meaning that the whole model still has to somehow maintain A-related information in the model while it's working with B. (This is no problem at all with RNNs cause RNN is inherently designed to fit this scenario.)

However convnet/transformer are not working that way.
As mentioned above I set useMask=true to make the model will produce the same output wrt A even though B follows subsequently.
But still for efficiency reason (i.e. genuine streaming), the model should keep the encoding of A as well as A itself in the model and proceed with the B.

Long story short, I wonder if I just give transformer as AM in your streaming decoder setting,
1. Will it seamlessly working fine?
If yes on 1,
2. Will it compute A and then A+B again for extracting B's encoding? or A and then B wisely?
3. Where should I look at the code to figure out above questions?

I wonder if I'm getting you lost, sorry for that :(

**Comment by lunixbochs** [comment]
The inference frontend doesn't implement all of the modules. Right now you can only convert roughly the modules used by sconv.

**Comment by mjhanphd** [comment]
OK I understood, could you do me a favor and briefly explain what makes those modules from sconv streaming friendly as opposed to others?

**Comment by mjhanphd** [comment]
Is the streaming implemented from the module? If so, what functionalities should I add to existing modules?

**Comment by mjhanphd** [comment]
From your [graphic illustration](https://ai.facebook.com/blog/online-speech-recognition-with-wav2letteranywhere/) of sconv, it's quite clear that the model processes input audio streams sequentially without repeatedly processing accumulated streams over and over.
So how is this possible especially for TDS as opposed to TR?
Seeing the implementation of each modules, I can't find any notable differences between them.
For both of them It's just the definition of the comprising primitive layers and forward functions.
How come only TDS is possible to process only the next incoming feature streams?

I'd really appreciate if you give me an answer.

**Comment by lunixbochs** [comment]
There are two problems:

1. Model Architecture
You should read the streaming convnet whitepaper: https://research.fb.com/publications/scaling-up-online-speech-recognition-using-convnets/

The 500ms forward context is the important part.

You can stream other architectures, but I've found they have very poor results unless you stream at least 1-2 second chunks, and generally have much better accuracy if you just give them the whole audio at once.

2. Inference Implementation
The code currently implemented in the wav2letter@anywhere inference frontend _cannot_ use any other architecture, because nobody has implemented the necessary modules in wav2letter's inference code to use any other model types, and the code in the streaming_tds_exporter does not support other architectures. If you or someone wrote the code (implemented the modules and improved the exporter), you could try to stream them with inference, but your model might not be very good.

For now, I recommend just using a variant of streaming convnets. It's a very good architecture.

**Comment by mjhanphd** [comment]
Thank you so much! Btw, what is exporter?

**Comment by mjhanphd** [comment]
I wonder if you could give me an excerpt of the code about it.

**Comment by mjhanphd** [comment]
I read the paper and also this one (https://arxiv.org/abs/2001.02674).
It also lets the transformer to look ahead to the future frames.
It shows great WER on Librispeech, so I wanted to test transformer using wav2letter. :)

**Comment by mjhanphd** [comment]
I'm also willing to code myself to better fit the transformer to the inference frontend.
Such as giving the transformer model an ability to attend to limited future context, and if needed and fix the module to support streaming.

Just for testing purpose I just eliminated all future attentions and tried training it with the train config of sconv, surprisingly it's showing better loss decline curve so far.

**Comment by lunixbochs** [comment]
Which transformer architecture? For me the transformer model sizes used by wav2letter recipes right now were far too slow to run on CPU.

This is the exporter: https://github.com/facebookresearch/wav2letter/blob/master/tools/StreamingTDSModelConverter.cpp

Please, please focus on reading more of the wiki and code in wav2letter instead of posting 4 responses in a row that could be answered by reading the documentation.

**Comment by vineelpratap** [comment]
Hi @mjhanphd, as mentioned above, we support only TDS based models as described in https://research.fb.com/publications/scaling-up-online-speech-recognition-using-convnets/

We are more than happy to accept contributions implementing more functionality into inference framework. Please check out https://github.com/facebookresearch/wav2letter/wiki/Inference-Framework for the details on how to the inference framework is implemented.
